### PR TITLE
Homework SHM #1

### PR DESCRIPTION
**Pull request by kszytko**
Authors:  @kszytko @KrzysztofJ85 @lksobczyk @Nessy90 @Tomonike

### CODE REVIEW

File: shm/Player.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    return this->_ship->getSpeed();
+}
+
+Cargo* Player::getCargo(size_t index) const {
+    return this->_ship->getCargo(index).get();
+}
+
+void Player::updateAvailableSpace() {
+    size_t cargoSize = std::accumulate(_ship->getCargo().begin(),
+                                       _ship->getCargo().end(),
```
Comment (kyrtaq) [inline shm/Player.cpp]: how does it work if getCargo needs an index to work?

Comment (kyrtaq) [reply shm/Player.cpp]: nvm, ive overlooked something

File: shm/Cargo.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+Cargo& Cargo::operator+=(const size_t amount) {
+    _amount += amount;
+    return *this;
+}
+
+Cargo& Cargo::operator-=(const size_t amount) {
+    _amount -= amount;
+    return *this;
+}
```
Comment (Vicx95) [inline shm/Cargo.cpp]: check that amount is not bigger than _amount :)

File: shm/Cargo.hpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <string>
+
+class Cargo {
+public:
+    Cargo(const std::string& name,
+          size_t amount,
+          size_t basePrice)
+        : _name(name), _amount(amount), _basePrice(basePrice) {}
```
Comment (Vicx95) [inline shm/Cargo.hpp]: move to cpp file :)

File: shm/Ship.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    size_t getSpeed() const { return _speed; }
+    size_t getMaxCrew() const { return _maxCrew; }
+    size_t getCapacity() const { return _capacity; }
+    size_t getCrew() const { return _crew; }
+    std::vector<std::shared_ptr<Cargo>> getCargo() const { return _cargo; }
+    std::shared_ptr<Cargo> getCargo(const size_t id) const {
+        if (id < _cargo.size()) {
+            return _cargo[id];
+        }
+        return {};
```
Comment (Vicx95) [inline shm/Ship.hpp]: move to cpp, because you have some logic here :)

File: shm/Ship.hpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
```
Comment (WBigaj) [inline shm/Ship.hpp:5]: Cargo.hpp has string included,so you do not need it here.

File: shm/Player.hpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+#include <memory>
```
Comment (WBigaj) [inline shm/Player.hpp:2]: Ship.hpp has memory,so you do not need it here.

File: shm/Cargo.cpp (hunk 2 above)
Comment (nauka-programowania-MA) [inline shm/Cargo.cpp]: Check if you don't reach max value or drop below 0

File: shm/Cargo.hpp (hunk 3 above)
Comment (nauka-programowania-MA) [inline shm/Cargo.hpp]: only definition of trivial getter/ setter could be in .hpp. Please move other to .cpp file.

File: shm/Map.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            _vecOfIslands.push_back(Island(positionX, positionY));
+        }
+    }
+}
+
+Island* Map::getIsland(const Coordinates& coordinate) {
+    auto itr = std::find_if(_vecOfIslands.begin(), _vecOfIslands.end(),
+                            [&](auto& i) {
+                                return i.getPosition() == coordinate;
+                            });
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp:31]: Check If you found result. In other case you got UB (undefined behavior).

File: shm/Player.cpp [hunk 8]
```diff
@@ -0,0 +1,26 @@
+#include "Player.hpp"
+
+#include <numeric>
+
+Player::Player(Ship ship, size_t money)
+    : _ship(std::make_unique<Ship>(ship)),
+      _money(money) {}
+
+size_t Player::getSpeed() const {
+    return this->_ship->getSpeed();
```
Comment (nauka-programowania-MA) [inline shm/Player.cpp]: this redundant

File: shm/Player.cpp [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+Cargo* Player::getCargo(size_t index) const {
+    return this->_ship->getCargo(index).get();
```
Comment (nauka-programowania-MA) [inline shm/Player.cpp]: this redundant

File: shm/Player.hpp (hunk 6 above)
Comment (nauka-programowania-MA) [reply shm/Player.hpp:2]: +1

File: shm/Ship.hpp (hunk 5 above)
Comment (nauka-programowania-MA) [reply shm/Ship.hpp:5]: +1

File: shm/Ship.hpp [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+         size_t speed,
+         size_t maxCrew);
+
+    int getId() const { return _id; }
+    std::string getName() const { return _name; }
+    size_t getSpeed() const { return _speed; }
+    size_t getMaxCrew() const { return _maxCrew; }
+    size_t getCapacity() const { return _capacity; }
+    size_t getCrew() const { return _crew; }
+    std::vector<std::shared_ptr<Cargo>> getCargo() const { return _cargo; }
```
Comment (nauka-programowania-MA) [inline shm/Ship.hpp]: Remove. We only want to give acces to Cargo* not whole vector

Comment (nauka-programowania-MA) [reply shm/Ship.hpp]: In other case you could break invariants.

File: shm/Ship.hpp (hunk 4 above)
Comment (nauka-programowania-MA) [reply shm/Ship.hpp]: +1

File: shm/main.cpp [hunk 11]
```diff
@@ -0,0 +1,19 @@
+#include <cassert>
+
+#include "Cargo.hpp"
+#include "Ship.hpp"
+
+int main() {
+    Ship ship1;
+    Ship ship2(10, 100, 500);
+    Ship ship3(10, "Ala", 200, 600, 300);
```
Comment (nauka-programowania-MA) [inline shm/main.cpp:9]: constexpr for this magic value :)

**Review by nauka-programowania-MA — CHANGES_REQUESTED** [review]
Nice work. Please fix issues and commit agian :)

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -0,0 +1,29 @@
+
```
Comment (kuchichan) [inline .github/workflows/main.yml:1]: Awesome!!!! 🚀 thanks for that, I wil try to implement in our project. Can you recommend some learning sources?

File: shm/Cargo.cpp [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+bool Cargo::operator==(const Cargo& src) const {
+     return (_name == src._name) && (_amount == src._amount) && (_basePrice == src._basePrice);
+}
+
+bool Cargo::operator!=(const Cargo& src) const {
```
Comment (kuchichan) [inline shm/Cargo.cpp:33]: Huge plus for !=operator

File: .github/workflows/main.yml (hunk 12 above)
Comment (kuchichan) [inline .github/workflows/main.yml:1]: + For ci, I will implement it in our project. Could you recommend some learning resources? :)

**Review by kuchichan — COMMENTED** [review]
GJ :)

File: .github/workflows/main.yml (hunk 12 above)
Comment (kszytko) [reply .github/workflows/main.yml:1]: I did it first time :P copy paste guy, uncle google and testing

File: shm/Cargo.cpp (hunk 2 above)
Comment (nauka-programowania-MA) [reply shm/Cargo.cpp]: +1

File: shm/Cargo.hpp (hunk 3 above)
Comment (nauka-programowania-MA) [reply shm/Cargo.hpp]: +1

File: shm/Map.cpp [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+    if (itr == _vecOfIslands.end())
+        return {};
```
Comment (nauka-programowania-MA) [inline shm/Map.cpp:34]: or return nullptr

File: shm/Map.hpp [hunk 15]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <vector>
+
+#include "Island.hpp"
+
+constexpr int minPositionXY = 1;
+constexpr int maxPositionXY = 10;
+constexpr int numberOfIslands = 10;
```
Comment (nauka-programowania-MA) [inline shm/Map.hpp:9]: This information is useful only for .cpp file.

**Review by nauka-programowania-MA — APPROVED** [review]

**Comment by nauka-programowania-MA** [comment]
{
"requirement_ids": [2568, 2568, 2568, 2568, 2568, 2568, 2568, 2568, 2569, 2570],
"points": [5, 5, 5, 5, 5, 5, 5, 5, 20, 6],
"users": ["kszytko", "KrzysztofJ85", "lksobczyk", "Nessy90", "Tomonike"]
}

**Pull request closed without merging**